packages/go_router/lib/src/route.dart

```diff
@@ -433,22 +460,347 @@ class ShellRoute extends RouteBase {
+  /// at any point in time. Each branch uses a separate [Navigator], identified
+  /// by [StatefulShellBranch.navigatorKey].
+  final StatefulShellBranchBuilder branchBuilder;
 
-  /// The [GlobalKey] to be used by the [Navigator] built for this route.
-  /// All ShellRoutes build a Navigator by default. Child GoRoutes
-  /// are placed onto this Navigator instead of the root Navigator.
+  /// Gets the state for the nearest stateful shell route in the Widget tree.
+  static StatefulShellRouteState of(BuildContext context) {
+    final InheritedStatefulNavigationShell? inherited = context
+        .dependOnInheritedWidgetOfExactType<InheritedStatefulNavigationShell>();
+    assert(inherited != null,
+        'No InheritedStatefulNavigationShell found in context');
+    return inherited!.routeState;
+  }
+}
+
+/// Representation of a separate navigation branch in a [StatefulShellRoute].
+///
+/// The only required argument is the rootLocation (or [rootLocations]), which
+/// identify the [defaultLocation] to be used when loading the branch for the
+/// first time (for instance when switching branch using the goBranch method in
+/// [StatefulShellBranchState]). The rootLocations also identify the valid root
+/// locations for a particular StatefulShellBranch, and thus on which Navigator
+/// those routes should be placed on.
+///
+/// A [navigatorKey] is optional, but can be useful to provide in case you need
+/// to use the [Navigator] created for this branch elsewhere.
+@immutable
+class StatefulShellBranch {
+  /// Constructs a [StatefulShellBranch].
+  StatefulShellBranch({
+    GlobalKey<NavigatorState>? navigatorKey,
+    List<String>? rootLocations,
+    String? rootLocation,
+    this.name,
+    this.restorationScopeId,
+    this.preload = false,
+  })  : assert(rootLocation != null || (rootLocations?.isNotEmpty ?? false)),
+        rootLocations = rootLocations ?? <String>[rootLocation!],
+        navigatorKey = navigatorKey ??
+            GlobalKey<NavigatorState>(
+                debugLabel: name != null ? 'Branch-$name' : null);
+
+  /// The [GlobalKey] to be used by the [Navigator] built for this branch.
+  ///
+  /// A separate Navigator will be built for each StatefulShellBranch in a
+  /// [StatefulShellRoute] and this key will be used to identify the Navigator.
+  /// The routes associated with this branch will be placed o onto that
+  /// Navigator instead of the root Navigator.
   final GlobalKey<NavigatorState> navigatorKey;
+
+  /// The valid root locations for this branch.
+  final List<String> rootLocations;
+
+  /// An optional name for this branch.
+  final String? name;
+
+  /// Whether this route branch should be preloaded when the associated
+  /// [StatefulShellRoute] is visited for the first time.
+  ///
+  /// If this is true, this branch will be preloaded by navigating to
+  /// the root location (first entry in [rootLocations]).
+  final bool preload;
```

Comment:
I think it shouldn't be a problem, since some apps require smooth transitions between branches

Reply:
This was mainly introduced to make it possible to have as smooth and flickerless transitions as possible when switching branches (as you say @ahmednfwela). But I see your point about possible misuse. We could always clarify in the documentation that care must be taken to minimise the work done at preload as much as possible. I think if people really want preload, they would find ways of doing it even if this flag wouldn't exist. And such workarounds would of course lead to additional complexity (and thus possibly more bugs and inefficiencies). But that's just speculation on my part, so I'll leave this decision to you.

Reply:
> Is there a use case that requires this? If this is purely a performance option, I suggest we don't expose this until this is a problem. This may actually lead to worse memory usage if develop abuse it.

Hi @chunhtai a use case for this would be to mimic the behavior of using `IndexedStacked` for nested navigation. When using `IndexedStacked` all the pages are preloaded even though they are not on the screen. A use case for this is a page that requires a network call, we want to give users a better experience by pre-loading that page before they even press on it so that when its time to view it, it will already have been loaded with the data instead of waiting for it to fetch only when needed. 

From my past experience, this can cause performance issues if not used properly, so yes developers have to keep in mind the initial performance cost of preloading screens but that is also depending on specific use cases. However, I guarantee you that if you remove this feature, a feature request will pop up in no time.

Reply:
What about renaming to be the opposite:
```dart
  /// Whether this route branch should be lazy-loaded once the route is visible to the user
  ///
  /// Defaults to true. If false, this branch will be preloaded when its associated [StatefulShellRoute] is 
  /// visited for  the first time.
  final bool lazy;
```

Reply:
sounds good to me. Alternatively we can add this flag on per StatefulBranch bases. Not that it would actually help with the underlying problem, but just throwing out some ideas. Either way, we should document the risk and warn developer to use it with care.

Reply:
The preload flag is/was defined per branch. I've updated the documentation now to clarify the purpose and cautioning against misuse. The field is still called preload though, but if lazy is considered a better name for this field, we can of course use that instead.